Mathlib/Algebra/Order/Archimedean/Class.lean

```diff
@@ -451,4 +461,185 @@ theorem liftOrderHom_mk (f : M → α) (h : ∀ a b, mk a ≤ mk b → f a ≤ f
 
 end LiftHom
 
+/-- Given a `UpperSet` of `MulArchimedeanClass`,
+all group elements belonging to these classes form a subsemigroup.
+This is not yet a subgroup because it doesn't contain the identity if s = ⊤. -/
+@[to_additive "Given a `UpperSet` of `ArchimedeanClass`,
+all group elements belonging to these classes form a subsemigroup.
+This is not yet a subgroup because it doesn't contain the identity if s = ⊤."]
+def subsemigroup (s : UpperSet (MulArchimedeanClass M)) : Subsemigroup M where
+  carrier := mk ⁻¹' s
+  mul_mem' {a b} ha hb := by
+    rw [Set.mem_preimage] at ha hb ⊢
+    obtain h | h := min_le_iff.mp (min_le_mk_mul a b)
+    · exact s.upper h ha
+    · exact s.upper h hb
+
+/-- Make `MulArchimedeanClass.subsemigroup` a subgroup by assigning
+s = ⊤ with a junk value ⊥. -/
+@[to_additive "Make `ArchimedeanClass.subsemigroup` a subgroup by assigning
+s = ⊤ with a junk value ⊥."]
+noncomputable
+def subgroup (s : UpperSet (MulArchimedeanClass M)) : Subgroup M :=
+  open Classical in
+  if hs : s = ⊤ then
+    ⊥
+  else {
+    subsemigroup s with
+    one_mem' := by
+      rw [subsemigroup, Set.mem_preimage]
+      obtain ⟨u, hu⟩ := UpperSet.coe_nonempty.mpr hs
+      simpa using s.upper (by simp) hu
+    inv_mem' := by simp [subsemigroup]
+  }
+
+variable {s : UpperSet (MulArchimedeanClass M)}
+
+@[to_additive]
+theorem subgroup_eq_of_ne_top (hs : s ≠ ⊤) : (subgroup s : Set M) = subsemigroup s := by
```

Comment:
```suggestion
theorem subgroup_eq_subsemigroup_of_ne_top (hs : s ≠ ⊤) : (subgroup s : Set M) = subsemigroup s := by
```
or
```suggestion
theorem subgroup_of_ne_top (hs : s ≠ ⊤) : (subgroup s : Set M) = subsemigroup s := by
```
or maybe even turn the lemma around? We are trying to encourage people to use `subgroup` over `subsemigroup`